src/i18n.ts

```diff
@@ -0,0 +1,280 @@
+    const { defaultLocale, localeNegotiator, useSession } = this.config;
+    return async function (
+      ctx: C & I18nContextFlavor,
+      next: NextFunction,
+    ): Promise<void> {
+      let translate: TranslateFunction;
+
+      ctx.i18n = {
+        fluent,
+        renegotiateLocale: negotiateLocale,
+        useLocale,
+        getLocale: getNegotiatedLocale,
+        setLocale,
+      };
+      ctx.t = translateWrapper;
+      ctx.translate = translateWrapper;
+
+      await negotiateLocale();
+      await next();
+
+      // Also exports ctx object properties for accessing them directly from
+      // the translation source files.
+      function translateWrapper(
+        key: string,
+        context?: TranslationContext,
+      ): string {
+        return translate(key, {
+          first_name: ctx.from?.first_name ?? "",
+          ctx: JSON.stringify(makeContextObject(ctx)),
+          ...context,
+        });
+      }
+
+      async function getNegotiatedLocale(): Promise<string> {
+        return await localeNegotiator?.(ctx) ??
+          // deno-lint-ignore no-explicit-any
+          (await (useSession && (ctx as any).session))?.__language_code ??
+          ctx.from?.language_code ??
+          defaultLocale;
+      }
+
+      // Determining the locale to use for translations
+      async function negotiateLocale(): Promise<void> {
+        const negotiatedLocale = await getNegotiatedLocale();
+        useLocale(negotiatedLocale);
+      }
+
+      function useLocale(locale: LocaleId): void {
+        translate = fluent.withLocale(locale);
+      }
+
+      async function setLocale(locale: LocaleId): Promise<void> {
+        if (!useSession) {
+          throw new Error(
+"You are calling `ctx.i18n.setLocale()` without setting `useSession` to `true` \
+in the configuration. It doesn't make sense because you cannot set a locale in \
+the session that way. When you call `ctx.i18n.setLocale()`, the bot tries to \
+store the user locale in the session storage. But since you don't have session \
+enabled, it cannot store the locale information in the session storage. You \
+should either enable sessions or use `ctx.i18n.useLocale()` instead.",
+          );
+        }
+
+        // deno-lint-ignore no-explicit-any
+        (await (ctx as any).session).__language_code = locale;
+        await negotiateLocale();
+      }
+    };
+  }
+}
+
+function makeContextObject(ctx: Context) {
+  const keys: Array<keyof Context> = [
+    "callbackQuery",
+    "channelPost",
+    "chatJoinRequest",
+    "chatMember",
+    "chat",
+    "chosenInlineResult",
+    "editedChannelPost",
+    "editedMessage",
+    "from",
+    "match",
+    "me",
+    "msg",
+    "message",
+    "inlineQuery",
+    "myChatMember",
+    "poll",
+    "pollAnswer",
+    "preCheckoutQuery",
+    "senderChat",
+    "shippingQuery",
+    "update",
+  ];
+
+  const obj: Record<string, unknown> = {};
+  for (const key of keys) obj[key] = ctx[key];
```

Comment:
No, I don't think you got the point here either. I just wanted to have all the getters of context in a JSON object, because I can't achieve them by doing this: `JSON.parse(ctx)`. So, there's no point in converting them to string, because we need the object, not a string version.

There could be a way to pass `ctx` as it is in the i18n translation context, but I wasn't able to do it yet. Maybe after the release, I'll look more into that too.